### ISSUE TITLE
Report use of MDX internals

### ISSUE THREAD

**Issue by remcohaszing** [mdx-js/mdx-analyzer#395]
### Initial checklist

- [X] I read the [support docs](https://mdxjs.com/community/support/)
- [X] I read the [contributing guide](https://mdxjs.com/community/contribute/)
- [X] I agree to follow the [code of conduct](https://github.com/mdx-js/.github/blob/HEAD/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Amdx-js&type=Issues) and couldn’t find anything (or linked relevant results below)

### Problem

MDX generates some internal variables in the compiled output. These variables are in scope and users _could_ use them, but they they _should not_.

### Solution

Report usage of the following identifiers:

- `_components`
- `_createMdxContent`

We should probably also report usage of the following variables:

- `MDXLayout`
- `MDXContent`

### Alternatives

Consider all variables starting with an underscore to be private.

**Comment by ChristianMurphy** [comment]
Good idea @remcohaszing!
Reporting on certain identifiers feels a bit like a lint rule 🤔
I'm curious on folks thoughts whether that should live: here, in the remark lint, or in eslint mdx.

**Comment by remcohaszing** [comment]
`props` allow a user to pass content from one part of code they own (the invocation of the MDX component) to another part of the code they own (the implementation of the MDX component). I think this is almost as useful as JSX expressions in MDX in general. Also other APIs may inject props, i.e. a HOC or Next.js data fetching API. Or a documentation writer could use variables defined by a website developer.

`_components` feels more like an implementation detail. It’s a convenient intermediate variable to support `props.components` and `useMdxComponents()`. Also lower case components are added on an as-needed basis, so it’s shape would unreliable for direct use.

`_createMdxContent` is an implementation detail to support MDX layout and React providers.

`_createMdxContent`, `MDXContent` would yield infinite recursion if used as JSX components inside MDX. Perhaps this is only a problem if this is used from within `MDXLayout` or `MDXContent`, not if used from other ESM exports.

Actually using `MDXLayout` could be fine. There’s not really a problem with wrapping nested parts with the same component that by coincidence also wraps the MDX content.

**Comment by remcohaszing** [comment]
> Reporting on certain identifiers feels a bit like a lint rule 🤔

Maybe, this is an interesting thought.

I was also thinking of filtering these names from autocompletion results.

**Comment by wooorm** [comment]
> Also lower case components are added on an as-needed basis, so it’s shape would unreliable for direct use.

(side note: the distinction isn’t about lower-case vs upper-case btw)

Agree that it’s probably a bad idea to use `_components.p` but someone could be accessing `_components.H1`. A merge of passed and injected components.

```mdx
An IIFE:

{
  (() => {
    // Access passed/injected component inside JS:
    return <_components.H1>Hi!</_components.H1>
  })()
}
```

If we want to allow this behavior, we should probably call it `components` and make sure it’s always defined.

> MDXContent would yield infinite recursion

If used in the main body. MDX files can also export components. Which could reference the global scope. Someone could do something along the lines of

```mdx
export function WrappedContent() {
  return <main><MDXContent /></main>
}

# Hi
```

**Comment by remcohaszing** [comment]
> If we want to allow this behavior, we should probably call it `components` and make sure it’s always defined.

I get your point, but if we do so, this:

```mdx
<HI />
```

is equivalent to:

```mdx
<components.HI />
```

It would only be different if the MDX file imports or defines the `HI` component, at which point it would be better to pick a different name for the locally defined component that doesn’t conflict with injected components IMO.

Also someone could already define the `components` property on `components`:

```jsx
<MDXContent components={{components: {HI}}} />
```

To introduce a new variable would be a breaking change.

**Comment by remcohaszing** [comment]
> > MDXContent would yield infinite recursion
> 
> If used in the main body. MDX files can also export components. Which could reference the global scope. Someone could do something along the lines of
> 
> ```mdx
> export function WrappedContent() {
>   return <main><MDXContent /></main>
> }
> 
> # Hi
> ```

Yep, usage from `mdxjsEsm` nodes is fine.

**Comment by wooorm** [comment]
Yes, renaming to make things public or preventing their use would both be breaking.

**Comment by remcohaszing** [comment]
On the compiler level, yes. To report bad usage in the editor is not breaking.

**Comment by remcohaszing** [comment]
Another internal variable is `_missingMdxReference`. This may or may not be defined in development mode.
